### ISSUE TITLE
Fix for broken ./build.sh (fails), need to add code to the /base/Dockerfile

### ISSUE THREAD

**Issue by BrianKayWalker** [brandon-rhodes/fopnp#16]
For the life of me I couldn't get this to work (need to enable virtualization in the laptop bios).  I later learned a lot more about box'es and vagrant (below is how I build the playground with 15.04 vivid64).

The instructions work 100%, but if you don't know what you are doing you might get messed up, so here is a detailed overview, hope it helps.

Install git, virtualbox and vagrant (`sudo apt-get install`  each of them).

Then follow the instructions:
`vagrant init brandon-rhodes/playground`     this command sets up the Vagrantfile with the commands to run

`vagrant up --provider virtualbox`      this command runs the Vagrantfile, the playground isn't found locally so it goes to the Atlas web site and downloads the brandon-rhodes/playground box, which is a preconfigured/bult Ubuntu 14.04 box running Docker with 12 containers that it networks to form a network playground where learners can experiment with HTTP, DNS, SSH, SMTP, and other popular network protocols.

`vagrant ssh`     this puts you into the VM you have access to all the playground code and files ( /home/vagrant/fopnp/ ) you can now run

`./launch.sh`       this launches the playground.

`ssh h1`   now you are in the playground docker machine h1 and can run commands

`traceroute www.example.com`   this runs a traceroute to the example web page.

Have Fun!
## note you have machines h1 through h4 per the documentation.

Building your own:  (you have access to all the files natively-- without being in the vagrant system)
Fortunately I have a super star son that worked with me and we figured it out.  First on the Ubuntu laptop go into the bios system configuration and enable virtualization technology. (Calvin -- IndyPy Fame,  found that issue).
To get the build.sh to run with vivid64 boxes, and then the launch.sh you need to go into the /fopnp/playground/base folder and edit the Dockerfile.  The last RUN apt-get install -y python3-crypto python3-lxml python3-zmq
add the following after the zmq libffi-dev libssl-dev
so it looks like 
RUN apt-get install -y python3-crypto python3-lxml python3-zmq libffi-dev libssl-dev
(you can use nano to edit the file).
I had to change the /fopnp/palyground/Vagrantfile  (again us nano to edit it).  The third line, config.vm.box = "ubunto/vivid64"  I had to change to config.vm.box = "vivid64"

I am guessing that depends on the box you installed with Vagrant.

So I am running a Ubuntu 16.04 and a 14.04, and did the same installs on both.  All is good.
1.  in Terminal (open it with Ctrl+Alt+t) cd to the directory you want to install to (you might need to create the directory with mkdir myinstall)
2. sudo apt-get install the following git, virtualbox, and vagrant
3. vagrant box add ubuntu/vivid64
4. sudo git clone  https://github.com/brandon-rhodes/fopnp.git
5. cd in to fopnp/playground/ and change the /fopnp/palyground/Vagrantfile  (again us nano to edit it).  The third line, config.vm.box = "ubunto/vivid64"  I had to change to config.vm.box = "vivid64"
6. go into the /fopnp/playground/base folder and edit the Dockerfile.  The last RUN apt-get install -y python3-crypto python3-lxml python3-zmq
add the following after the zmq,  libffi-dev libssl-dev
so it looks like 
RUN apt-get install -y python3-crypto python3-lxml python3-zmq libffi-dev libssl-dev
7. now cd to fopnp/playground
8. vagrant up
9. vagrant ssh
10.  ./build.sh  this takes a long time
11. ./launch.sh

Now you can ssh h1  then run the traceroute www.example.com

WOW you got it !!!!

**Comment by brandon-rhodes** [comment]
Thanks for the extended notes! I have linked to them from the main documentation so that others can benefit from them.

**Comment by tdarrow** [comment]
Can this be set up to run on ESXi?

**Comment by brandon-rhodes** [comment]
I don't know — alas, I don't have any experience with ESXi.

**Comment by Mohakhairy** [comment]
How can I run the Playground on Ubuntu 16.04 (No Virtualmachine/VirtualBox)

**Comment by brandon-rhodes** [comment]
@Modagiri — If you have Docker installed, it's possible you could skip down to the "Building the Playground" section of the README and follow the instructions for `./build.sh` and `./launch.sh`.

**Comment by upangka** [comment]
What's error???

vagrant@vagrant-ubuntu-trusty-64:~$ ./launch.sh    
++ dirname ./launch.sh                             
+ cd .                                             
+ grep -q docker                                   
++ groups                                          
+ echo vagrant                                     
+ sudo adduser vagrant docker                      
adduser: The group `docker' does not exist.

**Comment by janus** [comment]
I changed "from 14.04" to "from 16.04" and added the needed  libffi-dev and libssl-dev. I also changed the Vagrant file to xenial64. I was able to get started, I got build.sh and launch.sh up and running. I was able to do "ssh h[1-4]" but I have to manually copied ssh-config. But ssh www and their like failed. I noticed that they were no more included in the ssh-config file. But traceroute www.example.com also failed. 
I have done a quick check and I have noticed that you changed a couple of areas in previous commits. 
I noticed that "ip -a" ( works for hosts h[1-4])
My question is this, what can I do to get traceroute www.example.com to really work and "route n"

NB
vagrant@ubuntu-xenial:~/fopnp/playground$ brctl show
bridge name	bridge id		STP enabled	interfaces
docker0		8000.02421e9d94df	no		veth0ce2759
							veth7cc2652
							vetha20a496
							vetha6ca388
exampleCOM		8000.1ec262496bc4	no		example-eth1
							ftp-eth0
							www-eth0
homeA		8000.460183d56d1a	no		h1-eth1
							h2-eth1
							h3-eth1
							modemA-eth1
homeB		8000.36edb6133c8e	no		h4-eth1
							modemB-eth1

**Comment by janus** [comment]
@Q10Viking 
cd to playground and run ./build.sh there?
BTW what is the issue you faced?

**Comment by zseusz** [comment]
Brendon, thanks so much for your Playground - its on another level.
Please any one know how to install the playground box on a Windows 10 system running either VMware or VirtualBox. Much appreciated.

**Comment by brandon-rhodes** [comment]
For folks following this thread — someone pointed out to me recently that a few instructions were out of date in the various README files, so I've gone through and have made sure that none of the files still talks about the "play.sh" script, since it's been replaced with plain SSH as the way to connect to one of the playground hosts h1 through h4.

**Comment by rlefever** [comment]
Let me start by saying thanks for all the effort you have put into this. i am enjoying the book.

I Tried to set this up on Ubuntu 16.04 and had issues. Using vagrant I get up to vagrant up but it times out and I am never able to get further. The VM seems to be running and the 2222 port is there but it does not seem to be talking to anything on the VirtualBox side. unsure how to trouble shoot this.

Tried to do the stand alone docker version and again had some issues. Followed the advice above, had to use wzurowski/vivid64.

in DockerFile
Had to add 'RUN pip install -U pip' before the requirements2.txt ADD line.
Had to add 'RUN pip3 install -U pip' before the requirements.txt ADD line.

in build.sh and launch.sh had to comment out the adding vagrant to docker group
seems vagrant install via apt-get did not create the vagrant user and bootstrap.sh was not used.

Also had to add the h1 h2 h3 h4 to the /etc/hosts file. This caused ssh h1 to ssh into the localhost so had to ssh -p 2201 root@127.0.0.1 but finally got to the Ubuntu 14 hosts. Seems I have to put the password each time so I may have to manually add the the key to my trusted_keys file.

Last problem I am having is that h1 has eth0 172.17.0.x and eth1 192.168.1.11. I can ping to a couple of the 172 addresses but the 192's all fail as does traceroute and nslookup do I cannot tell if the other hosts are alive.

Any help would be greatly appreciated on getting either or both working. Thanks again.

**Comment by brandon-rhodes** [comment]
@rlefever Your problems remind me of the symptoms that I have seen when the IP addresses of part of a machine’s network are overlapping with the range of IP addresses that Docker is trying to use — only, of course, you are using raw VirtualBox and Vagrant. But a problem with connectivity would explain why Vagrant is timing out, if the packets were going to the wrong network interface?

**Comment by rlefever** [comment]
So, Windows 7 running VirtualBox.

Vagrant Host is a Ubuntu 16.04 VM running Vagrant and VirtualBox.
Docker Host is an Ubuntu 16.04 VM running Docker.

The Vagrant Host shows enp0s3 as 10.0.2.15/24 and the lo. network is set
to NAT in Main host VirtualBox settings.

The Docker host is showing docker0 as 172.17.0.1/16 and enp0s3 at
192.168.100.251/24 again a VM in VirtualBox, network set to Bridge.

The docker0 was assigned by the code I assume as I did not set it up.

changed Vagrant Host to Bridge and got the same results on vagrant up,
just times out.

changed the Docker host to NAT, The docker0 is still 172.168.0.1/16,
enp0s3 changed to 10.0.2.15/24. All the Docker images are using IPv6
addresses.

The 100.x/24 is my local network.

What are should I change to fix either/both?

Thanks

**Comment by brandon-rhodes** [comment]
Alas, I have never tried running with a vagrant host separate from the docker host. Without a couple of hours to sit down with your computer and explore which network connections and pings do and don’t work between the different hosts, I doubt if I could make any headway in figuring out how to get the different hosts talking successfully.

**Comment by rlefever** [comment]
I was hoping you could direct me to the files I needed to modify to
attempt making the corrections. Especially since I used the directions in
the README and the issue to get to this point. I find I learn best doing
troubleshooting.

Any guidance would be appreciated but I understand you cannot be expected
to maintain the playground forever and on every possible environment.

Any and all assistance/clues appreciated.

**Comment by rlefever** [comment]
Seems that where I am on the Vagrant box is that I cannot talk to the VM
to start the Docker containers. It seems the box.ovf says the vmname and
uuid are different that what are reported running by 'VBoxManage list
runningvms'.

The box.ovf says vmname is playground_<numbers>_<numbers>. The running one
is <user>_<numbers>_<numbers>.

**Comment by rlefever** [comment]
I think I have discovered the reason Option A (Vagrant Host) of the
playground would not work for me. It seems the VirtualBox VM requires
AMD-V support and I am running an Intel CPU. The VM therefore fails to
start is why the SSH connection never works. I cannot find a way around
that.

Any ideas on where to start to get networking operational on the Option B
(Docker Host)? I am not running boot2docker, any suggestions on what to
change in the launch.sh? Seems to me the place to start.

Thanks

**Comment by rlefever** [comment]
On the Option B docker only setup I have used

https://success.docker.com/article/troubleshooting-container-networking

to determine which hosts can or cannot talk to each other. Seems the
endpoints cannot talk to their respective 'router's (modemA, modemB, and
example). I have also found that the iptables -L command returns:

iptables v.1.8.3 (legacy): can't initialize iptables table 'filter':
Permission denied (you must be root)
Perhaps iptables or your kernel needs to be upgraded

When I issue whoami it returns root so I don't think that is the issue.

The host, Ubuntu 16.04, says the same message except v1.6.0 when I try to
iptables -L as a non-root user. Could it be a conflict of the iptables
version or is there a module not loading on the host or container side?

check if modules are loaded on the modemA container
lsmod | grep ip
ipt_MASQUERADE
iptable_nat
nf_conntrack_ipv4
nf_defrag_ipv4
nf_nat_ipv4
iptable_filter
ip_tables
x_tables
nf_nat
nf_conntrak
multipath

Any thoughts? My hopes are that you have this running and can/will compare.

Thanks.

**Comment by brandon-rhodes** [comment]
I’ve just gone into the `requirements.txt` files and added the exact
version numbers of each package that are installed in the official
playground image.  This means that folks building the playground
themselves (a) should no longer have errors because newer versions of
some packages require a more recent version of pip, and (b) that the
home-made playground they come out with should work the same inside as
the official playground image.

@zseusz — Apologies that I didn’t answer your question last year, I must
have missed it. Alas, I don’t use Windows, and so I’m not sure if
Windows machines require any extra steps beyond the usual `vagrant
init`, `vagrant up`, and `vagrant ssh` steps.

@rlefever — First, I have edited the README instructions for folks
building the playground themselves, so it correctly specifies the SSH
configuration file that will allow login to `h1` through `h4` without
running into any problems with hostnames or port numbers or needing a
password.  (In the official playground image that file gets installed as
the `~/.ssh/config` file itself, so no extra argument is needed for
folks running the official image.)

Second, I spent today launching the playground myself on a modern Ubuntu
box — your hope that “you have this running” is, alas, mistaken; I have
been using the official image for years and only today had time to try
out the old “build.sh” and “launch.sh” scripts on a modern machine.  I
found that modern Linux + Ubuntu + Docker no longer allows packets to
flow between across the virtual bridges that the script sets up between
the virtual hosts, so I have added “iptables” commands to “launch.sh”
that make the packets start flowing for me.  Try them out yourself, and
see if they make “ping” and other commands start working successfully!

Third, you typically can’t run iptables from inside the containers,
because the main host retains the right to impose iptables rules; always
run iptables from outside the various virtual hosts.

Good luck! Let me know what happens.
